teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,924 @@
+Task.
+
+#### Within a Pipeline Task
+
+Within the context of a Pipeline Task, the Pipeline author can refer to individual attributes of the object results of
+other Tasks.
+
+In the above example:
+
+```yaml
+value: "cloned $(tasks.cloned-git.results.cloned-gitrepo.url) at $(tasks.cloned-git.results.cloned-gitrepo.commitish)"
+```
+
+[As with params](#variable-replacement-with-object-params) only access to individual keys will be allowed initially.
+
+#### Collisions with builtin variable replacement
+
+Tekton provides builtin variable replacements which may one day conflict with the keys in a dictionary. Existing
+variable replacements will not conflict (the closest candidate is `path`, as in `$(results.foo.path)` but since this is
+only used in the context of writing Tasks results, there will be no conflict - see
+[variable replacement with object params](#variable-replacement-with-object-params) for more detail).
+
+But still, features we add in the future may conflict, for example if we added an optional params feature, we might
+provide a variable replacement like `$params.foo.bound`. What if the `foo` param was an object with a key called `bound` ?
+
+For example:
+
+```yaml
+results:
+  - name: foo
+    type: object
+    properties:
+      bound: {
+        type: string
+      }
+```
+
+[TEP-0080](https://github.com/tektoncd/community/blob/main/teps/0080-support-domainscoped-parameterresult-names.md#proposal)
+added support for using `[]` syntax instead of `.` to support variables which are have names that include `.`. Resolving
+the ambiguity here would only require using `$params.foo.bound` to refer to the built in variable replacement, and
+`$params.foo["bound"]` to unambiguously refer to the key `bound` within `foo`. Other options include:
+
+1. Do not allow keys to be defined which conflict with variable replacement
+    * Con: this means that a Task which is perfectly fine today might break tomorrow if we introduce new variable
+      replacement - in the above example, `foo.bound` is perfectly valid today and would be allowed but would suddenly
+      stop working after the `bound` feature is introduced
+2. Let the defined object override the variable replacement; i.e. in this case the built in replacement will not be
+   available
+
+## Test Plan
+
+In addition to unit tests the implementation would include:
+
+* At least one reconciler level integration test which includes an object param and an object result
+* A tested example of a Task, used in a Pipeline, which:
+    * Declares an object param
+    * Emits an object result
+    * In the Pipeline, another Task consumes a specific value from the object
+
+## Design Evaluation
+
+[See also the TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) design evaluation.
+
+* [Reusability](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability):
```

Comment:
Another improvement in reusability, if we have a schema, would be allowing for schemas to be reused and shared

Reply:
yes 💯  i think that would be a very likely next step after this